### PR TITLE
[FLINK-18887][python][connector/elasticsearch] Add Elasticsearch DataStream API

### PR DESCRIPTION
**Pull request by deadwind4**
## What is the purpose of the change

Add ElasticSearch connector for Python DataStream API

## Brief change log

  - *Add ElasticSearch6 connector for Python DataStream API*
  - *Add ElasticSearch7 connector for Python DataStream API*

## Verifying this change

This change added tests and can be verified as follows:

  - *Added test that validates Elasticsearch6SinkBuilder*
  - *Added test that validates Elasticsearch7SinkBuilder*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduces a new feature? (no)
  - If yes, how is the feature documented? (PyDocs)

### CODE REVIEW

File: flink-connectors/flink-connector-elasticsearch-base/src/main/java/org/apache/flink/connector/elasticsearch/PyTestElasticsearchEmitter.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+import org.apache.flink.connector.elasticsearch.sink.ElasticsearchEmitter;
+import org.apache.flink.connector.elasticsearch.sink.RequestIndexer;
+
+import org.elasticsearch.action.index.IndexRequest;
+import org.elasticsearch.client.Requests;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class PyTestElasticsearchEmitter implements ElasticsearchEmitter<String> {
```
Comment (deadwind4) [inline flink-connectors/flink-connector-elasticsearch-base/src/main/java/org/apache/flink/connector/elasticsearch/PyTestElasticsearchEmitter.java:13]: I don't know where this file is placed.

**Comment by deadwind4** [comment]
I find the ES has moved to the new repo. Whether do I write a doc in the new repo? @dianfu

**Comment by dianfu** [comment]
@deadwind4 Thanks a lot for the PR. It needs to be rebased to fix the conflicts.

**Comment by deadwind4** [comment]
I will write docs into this [repo](https://github.com/apache/flink-connector-elasticsearch).

File: flink-python/pyflink/datastream/connectors/elasticsearch.py [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+class ElasticsearchSinkBuilderBase(abc.ABC):
+    """
+    Base builder to construct a ElasticsearchSink.
+    """
+
+    @abc.abstractmethod
+    def __init__(self):
+        self._j_elasticsearch_sink_builder = None
+
+    def set_emitter(self, emitter_class_name: str):
```
Comment (dianfu) [inline flink-python/pyflink/datastream/connectors/elasticsearch.py:65]: This is not usable for Python users at all. There is no default implementations available for use. It means that Python users will have to write Java code to use ES connector.

File: flink-python/pyflink/datastream/connectors/elasticsearch.py [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+
+    @abc.abstractmethod
+    def get_http_host_class(self):
+        """
+        Gets the org.apache.http.HttpHost class which path is different in different Elasticsearch
+        version.
+        """
+        pass
+
+    def set_hosts(self, hosts: List[str]) -> 'ElasticsearchSinkBuilderBase':
```
Comment (dianfu) [inline flink-python/pyflink/datastream/connectors/elasticsearch.py:81]:
```suggestion
    def set_hosts(self, hosts: Union[str, List[str]]) -> 'ElasticsearchSinkBuilderBase':
```

File: flink-python/pyflink/datastream/connectors/elasticsearch.py [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+        return self
+
+    def set_connection_password(self, password: str) -> 'ElasticsearchSinkBuilderBase':
+        """
+        Sets the password used to authenticate the connection with the Elasticsearch cluster.
+        """
+        self._j_elasticsearch_sink_builder.setConnectionPassword(password)
+        return self
+
+    def set_connection_path_prefix(self, prefix: str):
```
Comment (dianfu) [inline flink-python/pyflink/datastream/connectors/elasticsearch.py:154]: Missing return result type hint? Please also check the other places.

File: flink-python/pyflink/datastream/connectors/__init__.py [hunk 5]
```diff
@@ -48,6 +50,10 @@
     'JdbcExecutionOptions',
     'NumberSequenceSource',
     'OutputFileConfig',
+    'ElasticsearchSink',
+    'Elasticsearch6SinkBuilder',
+    'Elasticsearch7SinkBuilder',
+    'FlushBackoffType',
```
[the remark this answers is not in the document]
Comment (deadwind4) [reply flink-python/pyflink/datastream/connectors/__init__.py]: > All the classes added in the future are placed in connector specific files
This makes sense to me. All classes in each connector keep unifying.

File: flink-python/pyflink/datastream/connectors/elasticsearch.py (hunk 2 above)
Comment (deadwind4) [reply flink-python/pyflink/datastream/connectors/elasticsearch.py]: I implement a `PythonSimpleElasticsearchEmiiter`, I have tested it insert and update is successful.

File: flink-connectors/flink-connector-elasticsearch-base/src/main/java/org/apache/flink/connector/elasticsearch/sink/PythonSimpleElasticsearchEmitter.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+package org.apache.flink.connector.elasticsearch.sink;
+
+import org.apache.flink.api.connector.sink2.SinkWriter;
+
+import org.elasticsearch.action.update.UpdateRequest;
+
+import java.util.Map;
+
+/** A ElasticsearchEmitter that is used Python Flink Connector. */
+public class PythonSimpleElasticsearchEmitter implements ElasticsearchEmitter<Map<String, Object>> {
```
Comment (dianfu) [inline flink-connectors/flink-connector-elasticsearch-base/src/main/java/org/apache/flink/connector/elasticsearch/sink/PythonSimpleElasticsearchEmitter.java:28]: It doesn't necessary to be used only in PyFlink and so we could remove the `Python` prefix.

File: flink-python/pyflink/datastream/connectors/elasticsearch.py [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+class ElasticsearchSinkBuilderBase(abc.ABC):
+    """
+    Base builder to construct a ElasticsearchSink.
+    """
+
+    @abc.abstractmethod
+    def __init__(self):
+        self._j_elasticsearch_sink_builder = None
+
+    def set_emitter(self, index: str, key: str, doc_type: str = None) \
```
Comment (dianfu) [inline flink-python/pyflink/datastream/connectors/elasticsearch.py:65]: What about introducing class `ElasticsearchEmitter` and adding a few utility methods:
- static_index(index: str, key_field: str=None, doc_type: str=None)
- dynamic_index(index_field: str, key_field: str=None, doc_type: str=None)

**Review by dianfu — COMMENTED** [review]
@deadwind4 Thanks for the PR. Have left a few comments. Besides, it would be great to also update the documentation: 
https://nightlies.apache.org/flink/flink-docs-master/docs/connectors/datastream/elasticsearch/

**Comment by deadwind4** [comment]
@flinkbot run azure

**Review by dianfu — APPROVED** [review]

**Pull request closed without merging**